vidyut-lipi/src/lib.rs

```diff
@@ -1,118 +1,551 @@
+    // valid encoding - but Rri is. So Rr is a valid prefix, so R, and Rr
+    // Need to be inserted as prefix but with no valid token translaton. 
+    fn add_prefixes(&mut self, sref: &mut String) {
+	while sref.chars().count() > 1 {
+	    // First check whether prefix already present.
+	    sref.pop();
+	    
+	    let cnode = self.enc_to_tok.get(sref);
+	    if cnode.is_none() {
+		let node = Node::init(sref.clone(), "".to_string(), "".to_string(),
+				      TOK_TRANSPARENT_TYPE, true, IMPLICIT_UNKNOWN);
+		self.insert_enc_to_tok(node);
+	    } else {
+		
+	    }
+	}
+    }
+    
+
+    fn get_enc_map(&self, sref: String) -> Option<&Node> {
+	let rval = self.enc_to_tok.get(&sref);
+	return rval;
+    }
+
+    #[allow(dead_code)]
+    fn get_tok_map(&self, sref: String) -> Option<&Node> {
+	return self.tok_to_enc.get(&sref)
+    }
+
+    pub fn update_string_with_token_with_implicit(&self, intok: String, instr: &mut String,
+						  imp: u8,prev:u8) {
+
+	let tnode = self.get_tok_map(intok);
+	if tnode.is_some() {
+	    let w = tnode.unwrap();
+	    match w.tok_type {
+		TOK_IS_CONSONANT => {
+		    instr.push_str(&w.mnem);
+		    if w.implicit != imp {
+			if imp == IMPLICIT_VIRAMA {
+			    let vnode = self.get_tok_map("TOK_VOWEL_SIGN_VIRAMA".to_string());
+			    if vnode.is_some() {
+				instr.push_str(&vnode.unwrap().mnem);
+			    }			    
+			}
+		    }
+		},
+		TOK_IS_VOWEL => {
+		    if prev == TOK_IS_CONSONANT {
+			    instr.pop(); // pop the associated virama
+			    instr.push_str(&w.assoc);
+		    } else {
+			instr.push_str(&w.mnem);
+		    }
+		}
+		TOK_LEXICAL_BREAK => {
+		    
+		},
+		_ => {
+		    instr.push_str(&w.mnem);
+		}
+	    }
+
+	    
+	    if w.implicit != imp && w.tok_type == TOK_IS_CONSONANT {
+		if imp == IMPLICIT_VIRAMA {
+
+		}
+	    }
+
+	    
+	}
+    }
+    
+    // debug print of the translationmap. 
+    pub fn print_transmap(&self) {
+	for (key, value) in &(self.enc_to_tok) {
+	    println!("Key: {}, {},{},{},{},{}", key, value.mnem,
+		     value.tok_val,
+		     value.suffix,
+		     value.tok_type,
+		     value.implicit);
+	}
+    }
+    
+}
+
+
+pub fn generate_map<P: AsRef<Path>>(path: P, transmap: &mut TranslationMap ) -> &mut TranslationMap {
+
+    let u  = read_encodings_from_file(path);
+    if u.is_err() {
+
+	match u {
+	    Err(e) => println!("Error processing json file {}", e),
+	    _ => println!("Unknown error ")
+	}
+
+	return transmap;
+    }
+
+    
+    let myscheme = u.unwrap();
+    // Maintain a current set of keys. that should already be in the hashmap.
+
+    let mut keys: Vec<String> = vec![];  
+
+    for e in myscheme.encodings {
+	let ttype: u8 = map_string_to_toktype(&e.ttype);
+	let implicit: u8 = map_string_to_implicit(&myscheme.implicit);
+	
+	
+        let mut indx = 0;
+
+	//println!("Enc - {:#?}", e);
+	
+	// for every string in enc insert an entry into the map.
+	
+        while indx < e.enc.len() {
```

Comment:
`for text in e.enc { ... }` ? No need for `indx` if you're simply accessing values in the vec.

Reply:
If you notice - the input string acts like a stack. If there is a way to push data back into an iterator i would use - but i couldn't find it.

Reply:
Hmm -- I don't follow. Here's how your loop seems to me:

```
while indx < e.enc.len() {
  let mnem = e.enc[indx].clone();
  [...]
  node.suffix = check_for_prefix(&keys, &e.enc[indx]);
  [...]
  let updmnemvec = check_existing_for_prefix_vec(&keys, &e.enc[indx]);
  [...]
  indx = indx + 1;
}
```

So we start at 0 and increment by one until `e.enc.len()`, and in each iteration, we get `e.enc[indx]`. I see pushes to `keys` but nothing pushed to `e.enc`. So I don't follow how this differs from a for loop.